UNL-CSE-AMBASSADORS/VITA
# Role selection

Removed the 'preparesTaxes' construct from the application, and replaced it with a system where a volunteer will select which role they are fulfilling on a day. This is more flexible than the old system.

**PRE-TESTING**
Here is the SQL script to generate the test data that I was using to verify the /getTimes functionality. Just run DB creation -> FullDBTestData -> this script. This will add 6 volunteers to shiftId 85, 3 of which are preparers, and the rest are not. Only the preparers should count toward the number of appointments we are allowed to make.
```
USE vita;

INSERT INTO User (firstName, lastName, email)
	VALUES ('Matthew', 'Meacham', 'matthew@test.test');
SET @user_matthew = LAST_INSERT_ID();

INSERT INTO User (firstName, lastName, email)
	VALUES ('Spencer', 'Collins', 'spencer@test.test');
SET @user_spencer = LAST_INSERT_ID();

INSERT INTO User (firstName, lastName, email)
	VALUES ('Austin', 'Schmidt', 'austin@test.test');
SET @user_austin = LAST_INSERT_ID();

INSERT INTO User (firstName, lastName, email)
	VALUES ('Jake', 'Koperski', 'jake@test.test');
SET @user_jake = LAST_INSERT_ID();

INSERT INTO User (firstName, lastName, email)
	VALUES ('Andre', 'Garivay', 'andre@test.test');
SET @user_andre = LAST_INSERT_ID();

INSERT INTO User (firstName, lastName, email)
	VALUES ('Rohan', 'Thakker', 'rohan@test.test');
SET @user_rohan = LAST_INSERT_ID();

SET @role_preparer = (SELECT roleId FROM Role WHERE lookupName = 'preparer');
SET @role_siteAdmin = (SELECT roleId FROM Role WHERE lookupName = 'site_administrator');
SET @role_greeter = (SELECT roleId FROM Role WHERE lookupName = 'greeter');
SET @role_reviewer = (SELECT roleId FROM Role WHERE lookupName = 'reviewer');
SET @shiftId = 85;
SET @siteId = (SELECT siteId FROM Shift WHERE shiftId = @shiftId);

INSERT INTO UserShift (userId, shiftId, roleId)
	VALUES (@user_matthew, @shiftId, @role_preparer);

INSERT INTO UserShift (userId, shiftId, roleId)
	VALUES (@user_spencer, @shiftId, @role_siteAdmin);

INSERT INTO UserShift(userId, shiftId, roleId)
	VALUES (@user_austin, @shiftId, @role_greeter);

INSERT INTO UserShift(userId, shiftId, roleId)
	VALUES (@user_jake, @shiftId, @role_reviewer);
    
INSERT INTO UserShift(userId, shiftId, roleId)
	VALUES (@user_andre, @shiftId, @role_preparer);
    
INSERT INTO UserShift(userId, shiftId, roleId)
	VALUES (@user_rohan, @shiftId, @role_preparer);
    
INSERT INTO AppointmentTime (scheduledTime, siteId)
	VALUES ('2018-03-28 16:30:00', @siteId);
INSERT INTO AppointmentTime (scheduledTime, siteId)
	VALUES ('2018-03-28 17:00:00', @siteId);
INSERT INTO AppointmentTime (scheduledTime, siteId)
	VALUES ('2018-03-28 17:30:00', @siteId);
INSERT INTO AppointmentTime (scheduledTime, siteId)
	VALUES ('2018-03-28 18:00:00', @siteId);
INSERT INTO AppointmentTime (scheduledTime, siteId)
	VALUES ('2018-03-28 18:30:00', @siteId);
```

**TESTING**
* Run DB refresher and build script
* Log in, go to profile, and sign up for a shift (note the role selection that is now there)
  * After signing up for a shift, it should show the role you selected in the signed up shift list
* Note the date that you signed up for, and navigate to /management/documents and download the volunteer schedule for that day
  * It should include the role you selected
* Navigate to /signup
* Ensure that the appointment sign up still functions as expected
* Navigate to /server/api/appointments/getTimes.php
  * Ensure results are as expected. The results will be less than last time
    * The expected results are the number of preparers working during that shift time. OR the maximum number of appointments.

**CHANGES TO PROD DB**
The following script needs to be ran against the PROD database WHEN this is deployed.
```
ALTER TABLE User DROP COLUMN preparesTaxes;

CREATE TABLE Role (
	roleId INTEGER UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
	name VARCHAR(255) NOT NULL,
	lookupName VARCHAR(255) NOT NULL,
	CONSTRAINT uniqueLookupName UNIQUE INDEX(lookupName)
);

ALTER TABLE UserShift ADD COLUMN roleId INTEGER UNSIGNED NOT NULL;
ALTER TABLE UserShift ADD CONSTRAINT fk_role_id FOREIGN KEY(roleId) REFERENCES Role(roleId);

INSERT INTO Role (name, lookupName)
	VALUES ("Site Administrator", "site_administrator");
INSERT INTO Role (name, lookupName)
	VALUES ("Greeter", "greeter");
INSERT INTO Role (name, lookupName)
	VALUES ("Preparer", "preparer");
INSERT INTO Role (name, lookupName)
	VALUES ("Reviewer", "reviewer");
```
